python_scripts/linear_models_feature_engineering_classification.py

```diff
@@ -0,0 +1,437 @@
+    x=feature_names[0],
+    y=feature_names[1],
+    hue=target_moons,
+    palette=["tab:red", "tab:blue"],
+    ax=axs[0],
+)
+sns.scatterplot(
+    data=gauss,
+    x=feature_names[0],
+    y=feature_names[1],
+    hue=target_gauss,
+    palette=["tab:red", "tab:blue"],
+    ax=axs[1],
+)
+sns.scatterplot(
+    data=xor,
+    x=feature_names[0],
+    y=feature_names[1],
+    hue=target_xor,
+    palette=["tab:red", "tab:blue"],
+    ax=axs[2],
+)
+axs[0].set_title("The moons dataset")
+_ = axs[1].set_title("The Gaussian quantiles dataset")
+_ = axs[2].set_title("The XOR dataset")
+
+
+# %% [markdown]
+#
+# We intuitively observe that for each dataset, the two classes are not
+# linearly separable as the classes fold around each other in a way no straight
+# line can separate the two classes.
+#
+# We can confirm this by fitting a linear model, such as logistic regression,
+# to each dataset and plot the decision boundary of the model. Let's first define
+# a helper function to plot the decision boundary of a model:
+
+# %%
+from sklearn.inspection import DecisionBoundaryDisplay
+
+
+def plot_decision_boundary(model):
+    _, axs = plt.subplots(ncols=3, figsize=(14, 4))
+
+    for ax, (data, target) in zip(
+        axs,
+        [
+            (data_moons, target_moons),
+            (data_gauss, target_gauss),
+            (data_xor, target_xor),
+        ],
+    ):
+        model.fit(data, target)
+        DecisionBoundaryDisplay.from_estimator(
+            model,
+            data,
+            response_method="predict_proba",
+            plot_method="pcolormesh",
+            cmap="RdBu",
+            alpha=0.8,
+            vmin=0,
+            vmax=1,
+            ax=ax,
+        )
+        DecisionBoundaryDisplay.from_estimator(
+            model,
+            data,
+            response_method="predict_proba",
+            plot_method="contour",
+            alpha=0.8,
+            levels=[0.5],
+            linestyles="--",
+            linewidths=2,
+            ax=ax,
+        )
+        sns.scatterplot(
+            data=data,
+            x=feature_names[0],
+            y=feature_names[1],
+            hue=target,
+            palette=["tab:red", "tab:blue"],
+            ax=ax,
+        )
+    return axs
+
+
+# %% [markdown]
+#
+# Now let's define our logistic regression model and plot its decision boundary
+# on the three datasets:
+
+# %%
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import StandardScaler
+from sklearn.linear_model import LogisticRegression
+
+logistic_regression = make_pipeline(StandardScaler(), LogisticRegression())
+logistic_regression
+
+# %%
+axs = plot_decision_boundary(logistic_regression)
+
+# %% [markdown]
+#
+# This confirms that it is not possible to separate the two classes with a
+# linear model. On each plot we see a **significant number of misclassified
+# samples on the training set**! This is a typical example of **underfitting**
+# for linear models.
```

Comment:
We can add a short comment on the somewhat unexpected behavior of the XOR plot, similar to "the last plot performs specially bad, as it tries to correctly classify most of the samples in the positive class, even if that means incorrectly identifying every sample of the negative class."

Reply:
Done in 0a53ccc.